### DISCUSSION TITLE
OpenSSL 3 Issues connecting to RMQ

### DISCUSSION THREAD

**Discussion by SpaceLineCrash** [rabbitmq/rabbitmq-server#9363] [Questions]
Hello RMQ,

I have a new environment using RabbitMQ 3.12.4 with Erlang 26. Our Ubuntu 22 clients have been unable to connect to this 3 node cluster due to an SSL error.  "ERROR [SSL: UNEXPECTED_EOF_WHILE_READING] EOF occurred in violation of protocol (_ssl.c:1007)". Clients are connecting through a HAProxy load balancer but we also attempted to connect directly which failed. 

The Ubuntu 20 clients do NOT have this issue above. The only major difference I am aware of is Openssl v1.1.1 on Ub20 to Openssl v3.0.2 on Ub22. From documentation, ERL 26 and RMQ 3.12 state they support Openssl V3. 

What configuration might I look into for my server or my clients to troubleshoot this? 

Haproxy Config:
frontend f-client1fronend-5671
  mode tcp
  maxconn 2500
  bind 1.1.1.4:5671 name 1.1.1.4:5671
  timeout client 500000
  tcp-request connection track-sc1 src
  default_backend b-upqueue2x2p-5671
  stick-table type ip size 1048576 expire 300000 store gpc0,conn_rate(1m),bytes_in_rate(10s),bytes_out_rate(10s),conn_cnt,sess_rate(10s)
backend b-clientbackendserver-5671
  mode tcp
  balance roundrobin
  timeout server 500000
  server clientbackendserver1_5671 1.1.1.1:5671 check fall 3 inter 2000 rise 5 weight 10
  server clientbackendserver2_5671 1.1.1.2:5671 check fall 3 inter 2000 rise 5 weight 10
  server clientbackendserver3_5671 1.1.1.3:5671 check fall 3 inter 2000 rise 5 weight 10

RabbitMQ Failing Client: Ubuntu 22, Openssl v3.0.2, Python3.10.12, pyOpenSSL 21.0.0

RabbitMQ Server Config below: 
# RabbitMQ General Settings
heartbeat          = 600
proxy_protocol     = true
log.file.level     = info
log.dir            = /var/log/rabbitmq
log.file           = rabbit.log
log.exchange       = true
log.exchange.level = info

# RabbitMQ Connection Settings
listeners.tcp                            = none
listeners.ssl.1                          = 5671
auth_mechanisms.1                        = EXTERNAL
auth_mechanisms.2                        = PLAIN
auth_mechanisms.3                        = AMQPLAIN
cluster_name                             = Client1-Prod
cluster_formation.peer_discovery_backend = dns
ssl_cert_login_from                      = common_name
ssl_options.cacertfile                   = /etc/ssl/lst/client1.cacert.pem
ssl_options.certfile                     = /etc/ssl/lst/client1.cert.pem
ssl_options.keyfile                      = /etc/ssl/lst/client1.key.pem
ssl_options.verify                       = verify_peer
ssl_options.versions.1                   = tlsv1.3
ssl_options.versions.2                   = tlsv1.2
ssl_options.ciphers.1                    = TLS_AES_128_GCM_SHA256
ssl_options.ciphers.2                    = TLS_AES_256_GCM_SHA384
ssl_options.ciphers.3                    = TLS_CHACHA20_POLY1305_SHA256
ssl_options.ciphers.4                    = ECDHE-ECDSA-AES128-GCM-SHA256
ssl_options.ciphers.5                    = ECDHE-RSA-AES128-GCM-SHA256
ssl_options.ciphers.6                    = ECDHE-ECDSA-AES256-GCM-SHA384
ssl_options.ciphers.7                    = ECDHE-RSA-AES256-GCM-SHA384
ssl_options.ciphers.8                    = ECDHE-ECDSA-CHACHA20-POLY1305
ssl_options.ciphers.9                    = ECDHE-RSA-CHACHA20-POLY1305
ssl_options.ciphers.10                   = DHE-RSA-AES128-GCM-SHA256
ssl_options.ciphers.11                   = DHE-RSA-AES256-GCM-SHA384
ssl_options.ciphers.12                   = DHE-RSA-CHACHA20-POLY1305
ssl_options.honor_cipher_order           = true
ssl_options.honor_ecc_order              = true
ssl_options.fail_if_no_peer_cert         = true

# RabbitMQ Management Settings
management.ssl.port       = 12345
management.ssl.versions.1 = tlsv1.2
management.ssl.cacertfile = /etc/ssl/lst/gdig2.crt
management.ssl.certfile   = /etc/ssl/lst/client1.pem
management.ssl.keyfile    = /etc/ssl/lst/client1.key

**Comment by michaelklishin** [comment]
RabbitMQ does not implement TLS or use OpenSSL (besides password hashing).

Your settings restrict supported versions and cipher suites. In addition to the [TLS troubleshooting methodology](https://rabbitmq.com/troubleshooting-ssl.html) suggested in the docs, try removing those
restrictions and re-introduce them one by one.

Specifically [TLS 1.3 support is an all-or-nothing proposition](https://rabbitmq.com/ssl.html#tls1.3)
because it by design does not share any cipher suites with earlier versions.

**Reply by SpaceLineCrash** [reply]
Hello Michael,

**This turned out to be a config issue.** I have the "proxy_protocol = true" however it was not enabled through HAProxy. I removed this line and I was able to connect in my environment. 

That being said my end goal is to have proxy_protocol enabled. Following the proxy documentation provided by HAProxy and RabbitMQ, linked below, my client is no longer able to connect. Would it be best to submit a new discussion or can it be addressed here? 

HAProxy Config:
frontend f-client1fronend-5671
mode tcp
maxconn 2500
bind 1.1.1.4:5671 name 1.1.1.4:5671 **accept-proxy**
timeout client 500000
tcp-request connection track-sc1 src
default_backend b-clientbackendserver-5671
stick-table type ip size 1048576 expire 300000 store gpc0,conn_rate(1m),bytes_in_rate(10s),bytes_out_rate(10s),conn_cnt,sess_rate(10s)

backend b-clientbackendserver-5671
mode tcp
balance roundrobin
timeout server 500000
server clientbackendserver1_5671 1.1.1.1:5671 check **send-proxy** fall 3 inter 2000 rise 5 weight 10   <--Also tried version 2 --> 
server clientbackendserver2_5671 1.1.1.2:5671 check **send-proxy** fall 3 inter 2000 rise 5 weight 10   <--Also tried version 2 --> 
server clientbackendserver3_5671 1.1.1.3:5671 check **send-proxy** fall 3 inter 2000 rise 5 weight 10   <--Also tried version 2 --> 


HAProxy/RabbitMQ Documentation:
https://www.haproxy.com/documentation/hapee/latest/load-balancing/client-ip-preservation/enable-proxy-protocol/
https://www.rabbitmq.com/networking.html#proxy-protocol

**Reply by michaelklishin** [reply]
I'm afraid our team does not have a time budget to assist with troubleshooting HAproxy installations.

If you use HAproxy then chances are you can use it for TLS termination and start without any TLS settings on the RabbitMQ end. Proxy protocol spec requires that no endpoint can support both Proxy protocol-enabled clients and non-enabled ones. So it must be enabled for RabbitMQ and for the proxy, other clients **will not** be able to connect (if any clients may bypass HAproxy, that is).

**Reply by SpaceLineCrash** [reply]
Hello Michael,

All clients connect through HAProxy so directly connected clients are not an issue for my use case.

That being said we use certificate peer verification for authentication on our RabbitMQ servers. Enabling TLS termination I thought would prohibit this verification process to work? 

I worked with HAProxy support to troubleshoot with some PCAPs. We can see that the proxy protocols are being sent from the client to HAProxy then rabbitMQ. That being said, the 3-way handshake works without the send-proxy protocol and fails as soon as it is enabled. The next course of action they provided me was to disable TLS to determine why send-proxy causes the RMQ handshake to fail. I couldn't get my client to connect without TLS enabled.

I also attempted to remove the restrictions I had on my ciphers and TLS versions but this reaped no benefit. 

Finally, I then troubleshot my TLS using the rabbitMQ page. I encountered an issue under the [Attempt TLS Connection to a RabbitMQ Node](https://www.rabbitmq.com/troubleshooting-ssl.html#sclient-connection) section again with send-proxy enabled. However, again when I disabled it, I received proper feedback from the TLS output.

BEFORE DISABLING SEND-PROXY:
```
root@rmqClientServer:/etc/ssl/# openssl s_client -connect clientqueue202p:5671 -cert /etc/ssl/rmqClientServer.cert.pem -key /etc/ssl/rmqClientServer.key.pem 
CONNECTED(00000003)
write:errno=104
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 293 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

AFTER DISABLING SEND-PROXY:
```
root@rmqClientServer:/etc/ssl/# openssl s_client -connect clientqueue202p:5671 -cert /etc/ssl/rmqClientServer.cert.pem -key /etc/ssl/rmqClientServer.key.pem
CONNECTED(00000003)
Can't use SSL_get_servername
depth=1 CN = ClientRabbitCA
verify error:num=19:self-signed certificate in certificate chain
verify return:1
depth=1 CN = ClientRabbitCA
verify return:1
depth=0 CN = clientqueue202p, O = server
verify return:1
---
Certificate chain
 0 s:CN = clientqueue202p, O = server
   i:CN = ClientRabbitCA
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: Aug  4 15:07:06 2023 GMT; NotAfter: Aug  1 15:07:06 2033 GMT
 1 s:CN = ClientRabbitCA
   i:CN = ClientRabbitCA
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: May  4 20:13:44 2016 GMT; NotAfter: May  2 20:13:44 2026 GMT
---
Server certificate
-----BEGIN CERTIFICATE-----
Omitted
-----END CERTIFICATE-----
subject=CN = clientqueue202p, O = server
issuer=CN = ClientRabbitCA
---
Acceptable client certificate CA names
CN = ClientRabbitCA
Requested Signature Algorithms: Omitted
Shared Requested Signature Algorithms: Omitted
Peer signing digest: Omitted
Peer signature type: Omitted
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 2114 bytes and written 1422 bytes
Verification error: self-signed certificate in certificate chain
---
New, TLSv1.3, Cipher is TLS_AES_128_GCM_SHA256
Server public key is 2048 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 19 (self-signed certificate in certificate chain)
---
closed
```

How would you recommend proceeding with troubleshooting? I would be happy to forward the PCAPs to you directly if they would be of benefit. 

Thank you,
Nate

**Reply by lukebakken** [reply]
@SpaceLineCrash 

> I have a new environment using RabbitMQ 3.12.4 with Erlang 26

Which packages are you using to install RabbitMQ and Erlang, and on what operating system?

Can you please provide a git repository that has the following in it:

* Complete Python test client code. All this has to do is open an AMQP connection using TLS / proxy protocol in the same way your real code is doing
* Complete HAproxy configuration
* Complete RabbitMQ configuration

I am asking for a git repo because then I know I'm using the same files as you.

What I can do is take the above information and set up an environment (probably using docker compose) to reproduce your issue. I am not going to set up a RabbitMQ cluster because that is not what is causing the issue here, I believe.

**Reply by SpaceLineCrash** [reply]
Hey Luke, we have a centralized HAProxy which carries most of our production environment. I will need some time to sanitize these items for you. Thank you for the reply and I will get this back to you shortly.

**Reply by SpaceLineCrash** [reply]
Hey Luke, I have a test consumer being created for me and I can update this ticket tomorrow with your requested information.

As an alternative and given this is already setup in my environment, would you be open to having a quick call to discuss/review configuration in our current environment? This would eliminate the need for you to build the vlab and we can do testing on my "production" environment.

**Reply by SpaceLineCrash** [reply]
Hello Luke, I added a private repo and added you and Michael. If there are any additional users, please let me know. 

https://github.com/SpaceLineCrash/RabbitMQ-TestBed.git

**Reply by lukebakken** [reply]
Thank you. I will investigate this as time allows.
